### PR TITLE
[Fix] Fixed behaviour for the conversion of complex tensors to bool

### LINKED ISSUE #120875: The result of torch.all is inconsistent between CPU and GPU
### 🐛 Describe the bug

The result of torch.all is inconsistent between CPU and GPU.

Codes:
```
import torch

input_data = torch.tensor([1+2j, 3-4j, 5j, 6])

result = torch.all(input_data)
print(result)

input_data = input_data.cuda()
result = torch.all(input_data)
print(result)
```

Outputs:
```
tensor(False)
tensor(True, device='cuda:0')
```

### Versions

Collecting environment information...
PyTorch version: 2.2.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.3 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.31

Python version: 3.8.18 (default, Sep 11 2023, 13:40:15)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-97-generic-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3090 Ti
Nvidia driver version: 535.154.05
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Byte Order:                         Little Endian
Address sizes:                      46 bits physical, 48 bits virtual
CPU(s):                             28
On-line CPU(s) list:                0-27
Thread(s) per core:                 2
Core(s) per socket:                 14
Socket(s):                          1
NUMA node(s):                       1
Vendor ID:                          GenuineIntel
CPU family:                         6
Model:                              85
Model name:                         Intel(R) Core(TM) i9-7940X CPU @ 3.10GHz
Stepping:                           4
CPU MHz:                            3100.000
CPU max MHz:                        4400.0000
CPU min MHz:                        1200.0000
BogoMIPS:                           6199.99
Virtualisation:                     VT-x
L1d cache:                          448 KiB
L1i cache:                          448 KiB
L2 cache:                           14 MiB
L3 cache:                           19.3 MiB
NUMA node0 CPU(s):                  0-27
Vulnerability Gather data sampling: Mitigation; Microcode
Vulnerability Itlb multihit:        KVM: Mitigation: VMX disabled
Vulnerability L1tf:                 Mitigation; PTE Inversion; VMX conditional cache flushes, SMT vulnerable
Vulnerability Mds:                  Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Meltdown:             Mitigation; PTI
Vulnerability Mmio stale data:      Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:             Mitigation; IBRS
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; IBRS, IBPB conditional, STIBP conditional, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Mitigation; Clear CPU buffers; SMT vulnerable
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single pti ssbd mba ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req md_clear flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.24.4
[pip3] torch==2.2.0
[pip3] triton==2.2.0
[conda] numpy                     1.24.4                   pypi_0    pypi
[conda] torch                     2.2.0                    pypi_0    pypi
[conda] triton                    2.2.0                    pypi_0    pypi

cc @ezyang @gchanan @zou3519 @kadeng @anjali411 @dylanbespalko @mruberry @Lezcano @nikitaved @amjames @albanD

### PR DESCRIPTION
**Pull request by andoorve**
Fixes #120875

### CODE REVIEW

File: aten/src/ATen/native/Copy.cpp [hunk 1]
```diff
@@ -296,8 +296,8 @@ static Tensor & copy_impl(Tensor & self, const Tensor & src, bool non_blocking)
   }
 #endif
 
-  if(!self.is_complex() && src.is_complex()) {
-    TORCH_WARN_ONCE("Casting complex values to real discards the imaginary part");
+if(!self.is_complex() && src.is_complex()) {
+    TORCH_WARN_ONCE("Casting complex values to real discards the imaginary part for non-bool casts");
```
Comment (lezcano) [inline aten/src/ATen/native/Copy.cpp:300]: Just throw this warning if `self` is not bool

File: c10/util/TypeCast.h [hunk 2]
```diff
@@ -48,12 +63,24 @@ template <typename dest_t, typename src_t>
...
     return static_cast<dest_t>(r);
   }
 };
 
+// Partial template specialization for casting to bool.
+// Need to handle complex types separately, as we don't 
+// simply want to cast the real part to bool. 
+template <typename src_t>
+struct static_cast_with_inter_type<bool, src_t> {
+  C10_HOST_DEVICE __ubsan_ignore_undefined__ static inline bool apply(
```
Comment (lezcano) [inline c10/util/TypeCast.h:77]: I don't think we have any UB wrapper here?

File: c10/util/TypeCast.h [hunk 3]
```diff
@@ -21,7 +21,7 @@ C10_CLANG_DIAGNOSTIC_IGNORE("-Wimplicit-int-float-conversion")
 namespace c10 {
 
 template <typename dest_t, typename src_t>
-struct needs_real {
+struct needs_real_or_bool {
```
Comment (lezcano) [inline c10/util/TypeCast.h:24]: Here "real" means "not complex", as `dest_t` may also be an int, so no need to change the name really.

**Review by lezcano — COMMENTED** [review]
Looking good. Could you add a regression test?

File: aten/src/ATen/native/Copy.cpp (hunk 1 above)
Comment (andoorve) [reply aten/src/ATen/native/Copy.cpp:300]: Great point 😅

File: c10/util/TypeCast.h (hunk 3 above)
Comment (andoorve) [reply c10/util/TypeCast.h:24]: Here I took the name to refer to "needs real from" `src_t` but let me change it back.

File: c10/util/TypeCast.h (hunk 2 above)
Comment (andoorve) [reply c10/util/TypeCast.h:77]: Yeah it's not necessary I think

**Comment by andoorve** [comment]
Added a couple of tests

File: test/test_complex.py [hunk 4]
```diff
@@ -33,6 +33,18 @@ def test_conj_copy(self, device, dtype):
         x1.copy_(xc1)
         self.assertEqual(x1, torch.tensor([5 - 1j, 2 - 2j], device=device, dtype=dtype))
 
+    @dtypes(*complex_types())
+    def test_all(self, device, dtype):
+        # issue: https://github.com/pytorch/pytorch/issues/120875
+        x = torch.tensor([1 + 2j, 3 - 4j, 5j, 6], device=device, dtype=dtype)
+        assert torch.all(x)
```
Comment (lezcano) [inline test/test_complex.py:40]:
```suggestion
       self.assertTrue(torch.all(x))
```

Comment (lezcano) [reply test/test_complex.py:40]: same below

**Review by lezcano — COMMENTED** [review]
Just a minor point. Otherwise lgtm.

And thank you for testing -0 and all its complex combinations!

File: test/test_complex.py (hunk 4 above)
Comment (andoorve) [reply test/test_complex.py]: Done!

**Review by lezcano — APPROVED** [review]
Thank you!

**Pull request closed without merging**